### ISSUE TITLE
ClientActor's message queues are sometimes very congested

### ISSUE THREAD

**Issue by marcelo-gonzalez** [near/nearcore#6438]
Recently we've seen a mainnet archival node built from b94948383f4106 process blocks too slowly to sync. What happens is that intermittently it'll fall behind and then stay around 500 blocks behind HEAD, as it doesn't process blocks fast enough to catch up after falling behind.

After adding logging like this:

<pre>diff --git a/chain/network/src/peer/peer_actor.rs b/chain/network/src/peer/peer_actor.rs
 index 3542deba4..103b1adbc 100644
 — a/chain/network/src/peer/peer_actor.rs
 +++ b/chain/network/src/peer/peer_actor.rs
 @@ -434,6 +434,8 @@ impl PeerActor 

 { NetworkClientMessages::PartialEncodedChunkRequest(request, msg_hash) } 

 RoutedMessageBody::PartialEncodedChunkResponse(response) => {
 + debug!(target: "network", "peer actor -> client actor chunk part {:?} {:?}",
 + &response.chunk_hash, response.parts.iter().map(|p| p.part_ord).collect::<vec>
 <u64>
 >());
 NetworkClientMessages::PartialEncodedChunkResponse(response)
 }
 </u64>
 </vec></pre>

<pre>diff --git a/chain/client/src/client_actor.rs b/chain/client/src/client_actor.rs
 index 0d2b1c792..032ee2ab7 100644
 — a/chain/client/src/client_actor.rs
 +++ b/chain/client/src/client_actor.rs
 @@ -564,6 +564,7 @@ impl Handler<networkclientmessages>
 for ClientActor 

 { NetworkClientResponses::NoResponse } 

 NetworkClientMessages::PartialEncodedChunkResponse(response) => {
 + debug!(target: "client", "recv partial chunk response {:?} {:?}", &response.chunk_hash, response.parts.iter().map(|p| p.part_ord).collect::
 <vec>
 <u64>
 >());
 if let Ok(accepted_blocks) =
 self.client.process_partial_encoded_chunk_response(response)
 {
 </u64>
 </vec>
 </networkclientmessages></pre>

We can see that there's big delay between when the peer actor passes the partial encoded chunk message to the client actor and the point when the client actor actually receives and processes it (sometimes seconds long):

 Mar 15 20:15:57.277 DEBUG network: peer actor -> client actor chunk part ChunkHash(`CooC7KRNfFsuUuEWaCAYeghCCujXZbm5LGUhEJmPWYm4`) <span class="error">[52]</span>
 ...
 ...
 Mar 15 20:16:00.799 DEBUG client: recv partial chunk response ChunkHash(`CooC7KRNfFsuUuEWaCAYeghCCujXZbm5LGUhEJmPWYm4`) <span class="error">[52]</span>

So the node ends up taking too long to sync even though there's nothing wrong with its block/chunk processing speed. Note that this is a node built from b94948383f4106 (plus the extra logging), which contains [https://github.com/near/nearcore/pull/6333](https://github.com/near/nearcore/pull/6333). So that PR definitely helps, but doesnt fully get rid of this

In debugging this, it could be helpful to add metrics that show this delay between when messages are sent to the client actor and when they're received (unless calculating it is too invasive/expensive with the volume of messages coming in, not sure).

cc @nikurt @bowenwang1996 @mm-near

**Comment by mm-near** [comment]
Yes - and the proper solution is to 'split' ClientActor into couple separate Actors/processes -- so that the runtime execution doesn't block it.

**Comment by bowenwang1996** [comment]
@nikurt did you assign yourself to fix the underlying issue or to add metrics?

**Comment by nikurt** [comment]
@bowenwang1996 to add metrics.
Let me unassign myself to avoid confusion

**Comment by marcelo-gonzalez** [comment]
@bowenwang1996 @nikurt ok filed https://github.com/near/nearcore/issues/6460
